devolutions-gateway/src/api/network_scan.rs

```diff
@@ -0,0 +1,123 @@
+        HttpError::internal().build(e)
+    })?;
+
+    let res = ws.on_upgrade(move |mut websocket| async move {
+        let stream = scanner.start().expect("Failed to start network scanner");
+        tracing::info!("Network scan started");
+        loop {
+            tokio::select! {
+                result = stream.recv() => {
+                    let Some((ip, dns, port)) = result else{
+                        break;
+                    };
+                    let response = NetworkScanResponse::new(ip, port, dns);
+                    let Ok(response) = serde_json::to_string(&response) else {
+                        warn!("Failed to serialize response");
+                        continue;
+                    };
+
+                    if let Err(e) = websocket.send(Message::Text(response)).await {
+                        tracing::info!("Failed to send message: {:?}", e);
+                        break;
+                    }
+
+                },
+                msg = websocket.recv() => {
+
+                    let Some(msg) = msg else {
+                        break;
+                    };
+
+                    if let Ok(Message::Close(_)) = msg {
+                        break;
+                    }
+                }
+            }
+        }
+        tracing::info!("Network scan finished");
+        stream.stop();
+    });
+    Ok(res)
+}
+
+#[derive(Debug, Deserialize)]
+pub struct NetworkScanQueryParams {
+    pub ping_interval: Option<u64>,     // in milliseconds,default 200
+    pub ping_timeout: Option<u64>,      // in milliseconds,default 500
+    pub broadcast_timeout: Option<u64>, // in milliseconds,default 1000
+    pub port_scan_timeout: Option<u64>, // in milliseconds,default 1000
+    pub netbios_timeout: Option<u64>,   // in milliseconds,default 1000
+    pub netbios_interval: Option<u64>,  // in milliseconds,default 200
+    pub max_wait: Option<u64>,          // max_wait for entire scan duration in milliseconds, suggested!
+}
+
+const FAMOUS_PORTS: [u16; 10] = [22, 23, 80, 443, 389, 636, 3389, 5900, 5985, 5986];
+impl From<NetworkScanQueryParams> for NetworkScannerParams {
+    fn from(val: NetworkScanQueryParams) -> Self {
+        NetworkScannerParams {
+            ports: FAMOUS_PORTS.to_vec(),
+            ping_interval: val.ping_interval.unwrap_or(200),
+            ping_timeout: val.ping_timeout.unwrap_or(500),
+            broadcast_timeout: val.broadcast_timeout.unwrap_or(1000),
+            port_scan_timeout: val.port_scan_timeout.unwrap_or(1000),
+            netbios_timeout: val.netbios_timeout.unwrap_or(1000),
+            max_wait_time: val.max_wait.unwrap_or(120 * 1000),
+            netbios_interval: val.netbios_interval.unwrap_or(200),
+        }
+    }
+}
+
+#[derive(Debug, Serialize)]
+pub struct NetworkScanResponse {
+    pub ip: String,
+    pub hostname: Option<String>,
+    #[serde(rename = "type")]
+    pub type_: String,
+}
```

Comment:
suggestion: For `ip`, you can use the `IpAddr` type directly instead of `String`. `Serialize` is implemented on it.

naming: `ty` is a common abbreviation to avoid the alien trailing `_`

BUT, question: what about calling this field `protocol` or `application_protocol` (we use the word "application protocol" abbreviated to "ap" in our tokens)

suggestion: Maybe we can reuse the `Protocol` type found in the `token` module, instead of `String`.
This way, we also keep the same stringified representation at both places (e.g.: "ssh" instead of "SSH" and the "winrm-http-pwsh" / "winrm-https-pwsh" combination instead of WinRM)

Reply:
updated